@log_capture() doesn't play nice with other decorators

Considering the following:


```
class MyTestTestCase(unittest.TestCase):

    @log_capture()
    def test_log_capture_only(self, log_capture):
        self.assertIsInstance(log_capture, LogCaptureForDecorator)

    @log_capture()
    @patch('apps.wideband.manager.order.OrderManager')
    def test_patch_and_log(self, mock, log_capture):
        self.assertIsInstance(log_capture, LogCaptureForDecorator)
        self.assertIsInstance(mock, MagicMock)

    @patch('apps.wideband.manager.order.OrderManager')
    @log_capture()
    def test_log_then_patch(self, log_capture, mock):
        self.assertIsInstance(log_capture, LogCaptureForDecorator)
        self.assertIsInstance(mock, MagicMock)
```
The following error is generated:

```
Error
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/testfixtures/utils.py", line 42, in wrapping
    return wrapped(*args, **kw)
  File "/usr/local/lib/python2.7/dist-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
TypeError: test_patch_and_log() takes exactly 3 arguments (2 given)

Failure
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/usr/local/lib/python2.7/dist-packages/testfixtures/utils.py", line 42, in wrapping
    return wrapped(*args, **kw)
  File "/opt/source/apps/wideband/tests/managers/test_order.py", line 106, in test_log_then_patch
    self.assertIsInstance(log_capture, LogCaptureForDecorator)
AssertionError: <MagicMock name='OrderManager' id='139896440938384'> is not an instance of <class 'testfixtures.logcapture.LogCaptureForDecorator'>)
```

This particular test case was tacked on to the end of a WIP file, so the line numbers won't quite match up.

The first test is fine.
The second test is not populating an argument for `log_capture`.
The third test actually has the order of the types swapped around.

When doing tests that have multiple patches, functionality has been that the `@patch` nearest the `def` appears first in the argument list. I would expect that the same applies here.

Using 
* Python 2.7
* Django 1.10.5
* mock 2.0.0
* testfixtures 5.4.0

--- a/testfixtures/tests/mock.py
+++ b/testfixtures/tests/mock.py
@@ -1,6 +1,6 @@
 from __future__ import absolute_import
 
 try:
-    from unittest.mock import Mock, call
+    from unittest.mock import Mock, call, patch
 except:
-    from mock import Mock, call
+    from mock import Mock, call, patch

--- a/testfixtures/tests/mock.py
+++ b/testfixtures/tests/mock.py
@@ -1,6 +1,6 @@
 from __future__ import absolute_import
 
 try:
-    from unittest.mock import Mock, call, patch
+    from unittest.mock import Mock, call, patch, MagicMock, DEFAULT
 except:
-    from mock import Mock, call, patch
+    from mock import Mock, call, patch, MagicMock, DEFAULT

--- a/testfixtures/tests/test_comparison.py
+++ b/testfixtures/tests/test_comparison.py
@@ -1,8 +1,10 @@
+from unittest import TestCase
+import sys
+
 from testfixtures import Comparison as C, TempDirectory, compare
 from testfixtures.compat import PY2, PY3, exception_module
 from testfixtures.tests.sample1 import SampleClassA, a_function
-from unittest import TestCase
-import sys
+import pytest
 
 
 class AClass:
@@ -47,6 +49,7 @@ class FussyDefineComparison(object):
         return not self == other  # pragma: no cover
 
 
+@pytest.mark.not_pytest
 class TestC(TestCase):
 
     def test_example(self):

--- a/testfixtures/tests/test_log_capture.py
+++ b/testfixtures/tests/test_log_capture.py
@@ -1,9 +1,11 @@
+from logging import getLogger
+from unittest import TestCase
+
+from mock import patch
+
 from testfixtures import (
     log_capture, compare, Comparison as C, LogCapture, ShouldRaise
 )
-from unittest import TestCase
-
-from logging import getLogger
 
 root = getLogger()
 one = getLogger('one')
@@ -210,6 +212,18 @@ class TestLog_Capture(TestCase):
             "('oh hai',)\n\n"
             "While comparing [0][2]: 'oh noez' (expected) != 'oh hai' (actual)"
         ))
+
+    @log_capture()
+    @patch('testfixtures.tests.sample1.SampleClassA')
+    def test_patch_then_log(self, a1, a2):
+        actual = [type(c).__name__ for c in (a1, a2)]
+        compare(actual, expected=['MagicMock', 'LogCaptureForDecorator'])
+
+    @patch('testfixtures.tests.sample1.SampleClassA')
+    @log_capture()
+    def test_log_then_patch(self, a1, a2):
+        actual = [type(c).__name__ for c in (a1, a2)]
+        compare(actual, expected=['LogCaptureForDecorator', 'MagicMock'])
 
 
 class BaseCaptureTest(TestCase):

--- a/testfixtures/tests/test_log_capture.py
+++ b/testfixtures/tests/test_log_capture.py
@@ -1,7 +1,9 @@
+from __future__ import absolute_import
+
 from logging import getLogger
 from unittest import TestCase
 
-from mock import patch
+from .mock import patch
 
 from testfixtures import (
     log_capture, compare, Comparison as C, LogCapture, ShouldRaise

--- a/testfixtures/tests/test_wrap.py
+++ b/testfixtures/tests/test_wrap.py
@@ -1,8 +1,8 @@
 from unittest import TestCase
 
-from .mock import Mock
+from .mock import Mock, MagicMock, patch, DEFAULT
 
-from testfixtures import wrap, compare
+from testfixtures import wrap, compare, log_capture
 
 
 class TestWrap(TestCase):
@@ -232,3 +232,14 @@ class TestWrap(TestCase):
             pass  # pragma: no cover
 
         compare(test_function.__name__, 'test_function')
+
+    def test_our_wrap_dealing_with_mock_patch(self):
+
+        @patch.multiple('testfixtures.tests.sample1', X=DEFAULT)
+        @log_capture()
+        def patched(log, X):
+            from testfixtures.tests.sample1 import X as imported_X
+            assert isinstance(X, MagicMock)
+            assert imported_X is X
+
+        patched()

--- a/testfixtures/utils.py
+++ b/testfixtures/utils.py
@@ -1,7 +1,27 @@
-from textwrap import dedent
-
+import sys
 from functools import wraps
 from inspect import getargspec
+from textwrap import dedent
+
+from . import singleton
+
+DEFAULT = singleton('DEFAULT')
+defaults = {DEFAULT}
+
+try:
+    from unittest.mock import DEFAULT
+except ImportError:
+    pass
+else:
+    defaults.add(DEFAULT)
+
+
+try:
+    from mock import DEFAULT
+except ImportError:
+    pass
+else:
+    defaults.add(DEFAULT)
 
 
 def generator(*args):
@@ -13,10 +33,25 @@ def generator(*args):
         yield i
 
 
-class Wrappings:
-    def __init__(self):
-        self.before = []
-        self.after = []
+def _is_started(patcher):
+    # XXXX horrible
+    return hasattr(patcher, 'is_local')
+
+
+class Wrapping:
+
+    attribute_name = None
+    new = DEFAULT
+
+    def __init__(self, before, after):
+        self.before, self.after = before, after
+
+    def __enter__(self):
+        return self.before()
+
+    def __exit__(self, exc_type=None, exc_val=None, exc_tb=None):
+        if self.after is not None:
+            self.after()
 
 
 def wrap(before, after=None):
@@ -24,34 +59,52 @@ def wrap(before, after=None):
     A decorator that causes the supplied callables to be called before
     or after the wrapped callable, as appropriate.
     """
-    def wrapper(wrapped):
-        if getattr(wrapped, '_wrappings', None) is None:
-            w = Wrappings()
 
-            @wraps(wrapped)
-            def wrapping(*args, **kw):
-                args = list(args)
-                to_add = len(getargspec(wrapped)[0][len(args):])
-                added = 0
-                for c in w.before:
-                    r = c()
-                    if added < to_add:
-                        args.append(r)
+    wrapping = Wrapping(before, after)
+
+    def wrapper(func):
+        if hasattr(func, 'patchings'):
+            func.patchings.append(wrapping)
+            return func
+
+        @wraps(func)
+        def patched(*args, **keywargs):
+            extra_args = []
+            entered_patchers = []
+
+            to_add = len(getargspec(func)[0][len(args):])
+            added = 0
+
+            exc_info = tuple()
+            try:
+                for patching in patched.patchings:
+                    arg = patching.__enter__()
+                    entered_patchers.append(patching)
+                    if patching.attribute_name is not None:
+                        keywargs.update(arg)
+                    elif patching.new in defaults and added < to_add:
+                        extra_args.append(arg)
                         added += 1
-                try:
-                    return wrapped(*args, **kw)
-                finally:
-                    for c in w.after:
-                        c()
-            f = wrapping
-            f._wrappings = w
-        else:
-            f = wrapped
-        w = f._wrappings
-        w.before.append(before)
-        if after is not None:
-            w.after.insert(0, after)
-        return f
+
+                args += tuple(extra_args)
+                return func(*args, **keywargs)
+            except:
+                if (patching not in entered_patchers and
+                    _is_started(patching)):
+                    # the patcher may have been started, but an exception
+                    # raised whilst entering one of its additional_patchers
+                    entered_patchers.append(patching)
+                # Pass the exception to __exit__
+                exc_info = sys.exc_info()
+                # re-raise the exception
+                raise
+            finally:
+                for patching in reversed(entered_patchers):
+                    patching.__exit__(*exc_info)
+
+        patched.patchings = [wrapping]
+        return patched
+
     return wrapper
 
 

--- a/testfixtures/utils.py
+++ b/testfixtures/utils.py
@@ -34,11 +34,6 @@ def generator(*args):
         yield i
 
 
-def _is_started(patcher):
-    # XXXX horrible
-    return hasattr(patcher, 'is_local')
-
-
 class Wrapping:
 
     attribute_name = None
@@ -90,11 +85,6 @@ def wrap(before, after=None):
                 args += tuple(extra_args)
                 return func(*args, **keywargs)
             except:
-                if (patching not in entered_patchers and
-                    _is_started(patching)):
-                    # the patcher may have been started, but an exception
-                    # raised whilst entering one of its additional_patchers
-                    entered_patchers.append(patching)
                 # Pass the exception to __exit__
                 exc_info = sys.exc_info()
                 # re-raise the exception

--- a/testfixtures/utils.py
+++ b/testfixtures/utils.py
@@ -8,8 +8,9 @@ from . import singleton
 DEFAULT = singleton('DEFAULT')
 defaults = {DEFAULT}
 
+
 try:
-    from unittest.mock import DEFAULT
+    from mock import DEFAULT
 except ImportError:
     pass
 else:
@@ -17,7 +18,7 @@ else:
 
 
 try:
-    from mock import DEFAULT
+    from unittest.mock import DEFAULT
 except ImportError:
     pass
 else:

--- a/testfixtures/utils.py
+++ b/testfixtures/utils.py
@@ -11,7 +11,7 @@ defaults = {DEFAULT}
 
 try:
     from mock import DEFAULT
-except ImportError:
+except ImportError:  # pragma: no cover
     pass
 else:
     defaults.add(DEFAULT)

